# KNN imputer

In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [15]:
df=pd.read_csv('framingham.csv')
print(df.shape)
df.head()

(4238, 16)


sex  age  education  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0    1   39        4.0              0         0.0     0.0                0   
1    0   46        2.0              0         0.0     0.0                0   
2    1   48        1.0              1        20.0     0.0                0   
3    0   61        3.0              1        30.0     0.0                0   
4    0   46        3.0              1        23.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             0         0    195.0  106.0   70.0  26.97       80.0     77.0   
1             0         0    250.0  121.0   81.0  28.73       95.0     76.0   
2             0         0    245.0  127.5   80.0  25.34       75.0     70.0   
3             1         0    225.0  150.0   95.0  28.58       65.0    103.0   
4             0         0    285.0  130.0   84.0  23.10       85.0     85.0   

   TenYearCHD  
0           0  
1           0  
2           0  
3           1  
4           0

# Exploratory Data Analysis

In [16]:
#df.info()

In [17]:
#df.describe()

In [18]:
df.isnull().sum()

sex                  0
age                  0
education          105
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

In [19]:
df.columns

Index(['sex', 'age', 'education', 'currentSmoker', 'cigsPerDay', 'BPMeds',
       'prevalentStroke', 'prevalentHyp', 'diabetes', 'totChol', 'sysBP',
       'diaBP', 'BMI', 'heartRate', 'glucose', 'TenYearCHD'],
      dtype='object')

In [7]:
#sns.pairplot(df)

In [8]:
#for assumption checking---normality
#sns.distplot(df['TenYearCHD'])

In [9]:
#import seaborn as sns
#fig,ax=plt.subplots(figsize=(14,12))
#sns.heatmap(df.corr(),annot=True)

# Fill out using KNN imputer

In [34]:
from sklearn.impute import KNNImputer

impute_knn=KNNImputer(n_neighbors=15); impute_knn
a=impute_knn.fit_transform(df); print(a)
df1=pd.DataFrame(a)
df1

[[  1.          39.           4.         ...  80.          77.
    0.        ]
 [  0.          46.           2.         ...  95.          76.
    0.        ]
 [  1.          48.           1.         ...  75.          70.
    0.        ]
 ...
 [  0.          48.           2.         ...  84.          86.
    0.        ]
 [  0.          44.           1.         ...  86.          78.26666667
    0.        ]
 [  0.          52.           2.         ...  80.         107.
    0.        ]]


0     1    2    3     4         5    6    7    8      9      10    11  \
0     1.0  39.0  4.0  0.0   0.0  0.000000  0.0  0.0  0.0  195.0  106.0  70.0   
1     0.0  46.0  2.0  0.0   0.0  0.000000  0.0  0.0  0.0  250.0  121.0  81.0   
2     1.0  48.0  1.0  1.0  20.0  0.000000  0.0  0.0  0.0  245.0  127.5  80.0   
3     0.0  61.0  3.0  1.0  30.0  0.000000  0.0  1.0  0.0  225.0  150.0  95.0   
4     0.0  46.0  3.0  1.0  23.0  0.000000  0.0  0.0  0.0  285.0  130.0  84.0   
...   ...   ...  ...  ...   ...       ...  ...  ...  ...    ...    ...   ...   
4233  1.0  50.0  1.0  1.0   1.0  0.000000  0.0  1.0  0.0  313.0  179.0  92.0   
4234  1.0  51.0  3.0  1.0  43.0  0.000000  0.0  0.0  0.0  207.0  126.5  80.0   
4235  0.0  48.0  2.0  1.0  20.0  0.066667  0.0  0.0  0.0  248.0  131.0  72.0   
4236  0.0  44.0  1.0  1.0  15.0  0.000000  0.0  0.0  0.0  210.0  126.5  87.0   
4237  0.0  52.0  2.0  0.0   0.0  0.000000  0.0  0.0  0.0  269.0  133.5  83.0   

         12    13          14   15  
0     26.97  80.0   77.000000  0.0  
1     28.73  95.0   76.000000  0.0  
2     25.34  75.0   70.000000  0.0  
3     28.58  65.0  103.000000  1.0  
4     23.10  85.0   85.000000  0.0  
...     ...   ...         ...  ...  
4233  25.97  66.0   86.000000  1.0  
4234  19.71  65.0   68.000000  0.0  
4235  22.00  84.0   86.000000  0.0  
4236  19.16  86.0   78.266667  0.0  
4237  21.47  80.0  107.000000  0.0  

[4238 rows x 16 columns]

In [21]:
df1.isnull().sum()

0     0
1     0
2     0
3     0
4     0
5     0
6     0
7     0
8     0
9     0
10    0
11    0
12    0
13    0
14    0
15    0
dtype: int64

In [35]:
y_=df1.iloc[:,[15]]
x_=df1.iloc[:,[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14]]
np.shape(x_)

(4238, 15)

In [36]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(x_,y_,test_size=0.3,random_state=101)
print(x_train.shape)
print(x_test.shape)

(2966, 15)
(1272, 15)


# Evaluating  using knn classification

In [37]:
from sklearn.neighbors import KNeighborsClassifier

knn=KNeighborsClassifier(n_neighbors=15)

In [38]:
knn.fit(x_train,y_train)

C:\Users\DeLL\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  """Entry point for launching an IPython kernel.


KNeighborsClassifier(n_neighbors=15)

In [39]:
#Predict the respose for test dataset

y_pred_=knn.predict(x_test)
y_pred_

array([0., 0., 0., ..., 0., 0., 0.])

In [40]:
#Confusion matrix

from sklearn.metrics import confusion_matrix

confMat=confusion_matrix(y_test,y_pred_)
confMat

array([[1087,   10],
       [ 167,    8]], dtype=int64)

In [41]:
#import scikit-learn metrics module for aaccuracy calculation
from sklearn import metrics

from sklearn.metrics import accuracy_score
#model accuracy

print("Accuracy:",metrics.accuracy_score(y_test,y_pred_))

Accuracy: 0.8608490566037735


Thus this model explained 86.08 % variability in 10 year risk of coronary heart disease CHD when missing values are replace using linear regression and logistic regression respective to that particular column and variable type.

# Evaluating  using logistic regression

In [42]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

lm=LogisticRegression(solver='liblinear',random_state=101).fit(x_train,y_train)

C:\Users\DeLL\Anaconda3\lib\site-packages\sklearn\utils\validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


In [43]:
print("R square : ", lm.score(x_train,y_train))

R square :  0.8506405933917734


Exception in callback BaseSelectorEventLoop._read_from_self()
handle: <Handle BaseSelectorEventLoop._read_from_self()>
Traceback (most recent call last):
  File "C:\Users\DeLL\Anaconda3\lib\asyncio\events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "C:\Users\DeLL\Anaconda3\lib\asyncio\selector_events.py", line 125, in _read_from_self
    data = self._ssock.recv(4096)
ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host
